Docs: Experiment best practices (#9860)
PostHog/posthog.com: contents/docs/experiments/best-practices.mdx

Author: Added "Test one change on one segment at a time. Multiple changes on the same segment makes it hard to know which has an impact."

Didnt add the other two. I dont think they're as useful to know, and would prefer to keep docs short and to the point

---

> title: Experiment best practices

Reviewer: Can keep the sidebar the same but this has 10x better SEO traffic.
```suggestion
title: A/B testing best practices
```

Reviewer 2: 100% this

Author: Sorry but disagree 🙃 We renamed it everywhere from "A/B tests" to "experiments" in the product and docs, so we should continue to refer to it as experiments here in the docs.

In any case, Im dropping the product name from the title and just calling it "Best practices" to keep consistent with other products - see [this comment](https://github.com/PostHog/posthog.com/pull/9858#discussion_r1840252050) on previous PR.

Reviewer: +1 on not dropping the name

sidebar: Docs
showTitle: true
---

> ## 1. Only include users affected by the change

Reviewer 3: ```suggestion
## 1. Target users precisely
```

Or some such. "the change" at the very top absent context feels jarring.

> Including users who aren't affected by the change being tested can lead to unreliable results. For example, if you're testing a new onboarding flow, you don't want to include users who have already completed the flow.

Reviewer 4: I wonder if "experiment" would be a better word here instead of "the change being tested"?

On the first time reading the sentence, "the change being tested" is a little hard to parse.

> Including users who aren't affected by the change being tested can lead to unreliable results. For example, if you're testing a new onboarding flow, you don't want to include users who have already completed the flow.

Author: ```suggestion
Including users who aren't affected by the experiment can lead to unreliable results. For example, if you're testing a new onboarding flow, you don't want to include users who have already completed the flow.
```

> To avoid including unaffected users, make sure to first filter out ineligible users in your code before checking your feature flag code. For example:

Reviewer 3: Oh,

> filter out ineligible users

could be a nice headline!

```js
// ✅ Correct. Will exclude unaffected users
function showNewChanges(user) {

  if (user.hasCompletedAction) {
    return false
  }

  // other checks

  if (posthog.getFeatureFlag('experiment-key') === 'control') {
    return false;
  }

  return true
}

// ❌ Incorrect. Will include unaffected users
function showNewChanges(user) {
  if (posthog.getFeatureFlag('experiment-key') === 'control') {
    return false;
  }

  if (user.hasCompletedAction) {
    return false
  }

  // other checks

  return true
}
```

## 2. Test your experiment

> Sometimes we're so eager to get results from our experiments, we jump straight to running them with all our users. This may be okay if everything is set up correctly, but if you've made a mistake, you may be unable to rerun your experiment. To understand why, let's look at an example:

Reviewer: ```suggestion
Sometimes we're so eager to get results from our experiments, we jump straight to running them with all our users. This is okay if everything is set up correctly, but if you've made a mistake, you may be unable to rerun your experiment. To understand why, let's look at an example:
```

> Imagine you're running an experiment with a 50/50 split between control and test. You roll out the experiment to all your users, but a day after launch, you notice that your change is causing the app to crash for all test users. You immediately stop the experiment and fix the root cause of the crash. However, restarting the experiment now will produce unreliable results since many users have already seen your change.

Reviewer 3: nice

> To avoid this problem, you should first test your experiment with a [small rollout](/tutorials/phased-rollout) (e.g., 5% of users) for a few days. Once you're confident everything works correctly, you can start the experiment with the remaining users.

Reviewer: How does someone do this in PostHog? Should probably call out they need to edit the feature flag before launching because it is non-obvious.

Here is a list of what to check during your test rollout:
- Logging is working correctly
- No increase in crashes or other errors
- Use [session replays](/session-replay) to ensure your app is behaving as expected
- Users are assigned to the control and test groups in the ratio you are expecting (e.g., 50/50).

> ## 3. Establish a clear hypothesis

Reviewer: This should probably go #1, because it is the first thing to think about when running an experiment.

A good hypothesis focuses your testing and guides your decision-making. It should include your goal metric, how you think your change will improve it, and any other important context. For example:

> 

Reviewer: ```suggestion
```

> Showing a short tutorial video during onboarding will help users understand how to use our product. As a result, we expect to see more successful interactions with the app, fewer customer support queries, and reduced churn (our primary goal) among tested users.

This is a good hypothesis because it enables you to break down which metrics you'll focus on in your test:

- How many users watch the video

> - What portion of the video did they view

Reviewer 3: ```suggestion
- What portion of the video they viewed
```

Alternatively, if you don't have enough statistical power (i.e., not enough users to obtain a significant result), you'll potentially waste weeks waiting for results. This is especially common in [group-targeted experiments](/blog/running-group-targeted-ab-tests).

For these reasons, PostHog includes a recommended running time calculator in the experiment setup flow. This calculates the minimum sample size required to run your experiment and the duration you should run your experiment for.

<ProductVideo
    videoLight= "https://res.cloudinary.com/dmukukwp6/video/upload/experiment_light_7070cd279b.mp4" 
    videoDark= "https://res.cloudinary.com/dmukukwp6/video/upload/experiment0dark_61a5bcbb63.mp4"
    alt="How to use the A/B test duration calculator" 
    classes="rounded"
/>

## 5. Use a launch checklist

> Launching an A/B test requires careful planning to ensure accurate results and meaningful insights. To help you navigate this, we've put together this launch checklist:

Reviewer 3: so good!

---

> title: Best practices

Author: ```suggestion
title: Experiments & A/B testing best practices
```
